src/trackinfoobject.cpp

```diff
@@ -861,64 +832,61 @@ void TrackInfoObject::slotCueUpdated() {
     emit(cuesUpdated());
 }
 
-Cue* TrackInfoObject::addCue() {
+CuePointer TrackInfoObject::addCue() {
     //qDebug() << "TrackInfoObject::addCue()";
     QMutexLocker lock(&m_qMutex);
-    Cue* cue = new Cue(m_iId);
-    connect(cue, SIGNAL(updated()),
+    CuePointer pCue(new Cue(m_iId));
+    connect(pCue.data(), SIGNAL(updated()),
             this, SLOT(slotCueUpdated()));
-    m_cuePoints.push_back(cue);
-    setDirty(true);
-    lock.unlock();
+    m_cuePoints.push_back(pCue);
+    setDirtyAndUnlock(&lock, true);
     emit(cuesUpdated());
-    return cue;
+    return pCue;
 }
 
-void TrackInfoObject::removeCue(Cue* cue) {
+void TrackInfoObject::removeCue(CuePointer pCue) {
     QMutexLocker lock(&m_qMutex);
-    disconnect(cue, 0, this, 0);
-    // TODO(XXX): Delete the cue point.
-    m_cuePoints.removeOne(cue);
-    setDirty(true);
-    lock.unlock();
+    disconnect(pCue.data(), 0, this, 0);
+    m_cuePoints.removeOne(pCue);
+    setDirtyAndUnlock(&lock, true);
     emit(cuesUpdated());
 }
 
-const QList<Cue*>& TrackInfoObject::getCuePoints() {
+QList<CuePointer> TrackInfoObject::getCuePoints() const {
     QMutexLocker lock(&m_qMutex);
     return m_cuePoints;
 }
 
-void TrackInfoObject::setCuePoints(QList<Cue*> cuePoints) {
+void TrackInfoObject::setCuePoints(QList<CuePointer> /*by-value*/ cuePoints) {
```

Comment:
Qt has a feature called implicit sharing. All standard containers types are thin-wrappers around a reference counted copy-on-write interior data structure. So passing a QList by value is copying a pointer and incrementing a reference count. 

Mutating the QList with swap actually calls ```QList::detach()``` which in the general case copies the structure. So I think it'd be better to just continue using assignment and not swap since it's free until someone mutates the container.